rpcwebsocket.go

```diff
@@ -242,6 +245,204 @@ func (m *wsNotificationManager) NotifyMempoolTx(tx *btcutil.Tx, isNew bool) {
 	}
 }
 
+// wsClientFilter tracks relevant addresses for each websocket client for
+// the `rescanblocks` extension. It is modified by the `loadtxfilter` command.
+//
+// NOTE: This extension was ported from github.com/decred/dcrd
+type wsClientFilter struct {
+	mu sync.Mutex
+
+	// Implemented fast paths for address lookup.
+	pubKeyHashes        map[[ripemd160.Size]byte]struct{}
+	scriptHashes        map[[ripemd160.Size]byte]struct{}
+	compressedPubKeys   map[[33]byte]struct{}
+	uncompressedPubKeys map[[65]byte]struct{}
+
+	// A fallback address lookup map in case a fast path doesn't exist.
+	// Only exists for completeness.  If using this shows up in a profile,
+	// there's a good chance a fast path should be added.
+	otherAddresses map[string]struct{}
+
+	// Outpoints of unspent outputs.
+	unspent map[wire.OutPoint]struct{}
+}
+
+// makeWSClientFilter creates a new, empty wsClientFilter struct to be used
+// for a websocket client.
+//
+// NOTE: This extension was ported from github.com/decred/dcrd
+func makeWSClientFilter(addresses []string, unspentOutPoints []*wire.OutPoint) *wsClientFilter {
```

Comment:
the make vs new distinction is not for pointers vs non-pointers.  it's for pointers to zero allocations vs non-zero initialized data